### PR TITLE
Add catalog queries to `lksearch`

### PR DESCRIPTION
**Pull request by christinahedges**
`lksearch` is our package for searching and finding data. These queries all require the internet to work. It seems that the logical place for catalog queries to live is inside `lksearch`. 

Catalog queries are useful for users, and can be agnostic of any particular file. @rebekah9969 has already opened a PR against `lightkurve` with this functionality, I have fixed the remaining issues with the PR and changed it to output a `pandas.DataFrame` object, to be more inline with the rest of `lksearch`.

Here's an example of the functional call with output

Screenshot 2024-07-29 at 2 41 24 PM

There are convenience functions to query each individual catalog by name and it will let you query by pixel distance.

```python
from lksearch import query_EPIC
catalog = query_EPIC(SkyCoord.from_name('Kepler-10'),
              Time.now(),
              radius=2*u.pixel,
              magnitude_limit=18)
```

### CODE REVIEW

File: src/lksearch/__init__.py [hunk 1]
```diff
@@ -78,8 +79,10 @@ class Conf(_config.ConfigNamespace):
 
 
 conf = Conf()
+log = logging.getLogger("lksearch")
 
-from .MASTSearch import MASTSearch
-from .TESSSearch import TESSSearch
-from .KeplerSearch import KeplerSearch
-from .K2Search import K2Search
+from .MASTSearch import MASTSearch  # noqa
```
Comment (Nschanche) [inline src/lksearch/__init__.py:84]: What is noqa?

File: src/lksearch/catalog.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        astropy.table.Table which contains the coordinates of targets and proper motion values
+    equinox: astropy.time.Time
+        The equinox for the catalog
+    epoch : astropy.time.Time
+        Time of the observation - This is taken from the table R.A and Dec. values and re-formatted as an astropy.time.Time object
+
+    Returns:
+    ------
+    table : astropy.table.Table
+        Returns an astropy table with ID, corrected RA, corrected Dec, and Mag(?Some ppl might find this benifical for contamination reasons?)
```
Comment (Nschanche) [inline src/lksearch/catalog.py:74]: I agree it is useful. Do all tables have the same reference year?

File: src/lksearch/catalog.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    """
+
+    # We need to remove any nan values from our proper  motion list
+    # Doing this will allow objects which do not have proper motion to still be displayed
+    table["pmRA"] = np.ma.filled(table["pmRA"].astype(float), 0.0)
```
Comment (Nschanche) [inline src/lksearch/catalog.py:79]: Ugh, masked arrays, ok.

File: src/lksearch/catalog.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    """
+    Returns an astropy table of sources with the proper motion correction applied
+
+    Parameters:
+    -----------
+    table :
+        astropy.table.Table which contains the coordinates of targets and proper motion values
+    equinox: astropy.time.Time
+        The equinox for the catalog
+    epoch : astropy.time.Time
```
Comment (Nschanche) [inline src/lksearch/catalog.py:68]: How will this be called in the workflow? Catalog values won't have a relevant epoch, so is this expected to be called from a lksearch result for actual data?

File: src/lksearch/catalog.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+    pmzero_index = np.where((table["pmRA"] == 0.0) & (table["pmDEC"] == 0.0))
+
+    # In those instances replace with J2000 values
+    table["RA"][pmzero_index] = table["RAJ2000"][pmzero_index]
+    table["DEC"][pmzero_index] = table["DEJ2000"][pmzero_index]
+    return table
+
+
+def query_catalog(
+    coord: SkyCoord,
```
Comment (Nschanche) [inline src/lksearch/catalog.py:122]: Should this allow a name and resolve the coordinates?

File: src/lksearch/catalog.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+    sep = coord.separation(c1)
+
+    # Find the object with the minimum separation - this is our target
+    ref_index = np.argmin(sep)
+
+    # apply_propermotion
+    result = _apply_propermotion(result, equinox=catalog_meta["equinox"], epoch=epoch)
+
+    # Now we want to repete but using the values corrected for proper motion
```
Comment (Nschanche) [inline src/lksearch/catalog.py:232]: repeat

File: src/lksearch/catalog.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+    return query_catalog(
+        coord=coord,
+        epoch=epoch,
+        catalog="kic",
+        radius=radius,
+        magnitude_limit=magnitude_limit,
+    )
+
+
+def query_TIC(
```
Comment (Nschanche) [inline src/lksearch/catalog.py:296]: Do we want to have query functions for each catalog, or as a parameter in the main call, like query_catalog(catalog='TIC')

File: src/lksearch/catalog.py (hunk 4 above)
Comment (Nschanche) [reply src/lksearch/catalog.py:68]: As this is passed from epoch in the query calls, I would be more explicit in the documentation that this time is used for PM corrections. I would also suggest this is optional.  Right now I think how 'epoch' is set in this is confusing.

File: src/lksearch/catalog.py (hunk 2 above)
Comment (christinahedges) [reply src/lksearch/catalog.py]: No they don't, gaiaedr3 has a different year. It's tracked in the dictionary.

File: src/lksearch/catalog.py (hunk 4 above)
Comment (christinahedges) [reply src/lksearch/catalog.py:72]: I think users need this functionality, but I've set it to default to the catalogs equinox.

**Comment by christinahedges** [comment]
@Nschanche thanks for a great review, I've updated this PR based on your comments, and a few more tweaks. 

- I updated epoch to have a sane default
- I changed the outputs to remove the "motion" columns. This was on purpose because the returned table has already been updated for motion, so I didn't want users to be confused over which RA/Dec to use
- I added the ability to get the skycoord returned instead of a table, which will be helpful for some workflows.
- I added the name resolve so if users put in "Au mic" it will resolve inside the query. 
- I added a tutorial. To do this I had to make a new landing page so all the tutorials are accessible. 

@tylerapritchard could you give this a review?

@Nschanche and @tylerapritchard are you happy with this API? An alternate would be:

```
from lksearch.catalogs import TICSearch
c = SkyCoord.from_name('Kepler-10')
sr = TICSearch(c)
sr.table
sr.skycoords
```

But I think this might be confusing because there's no e.g. `download` function, and no need to `filter` anything. I think this API is simpler for people, but what do you think?

**Comment by tylerapritchard** [comment]
This looks great - I'll review the code - but from a high level what if we renamed query_catalog to CatalogSearch to keep the naming theme but not change the API?  I think this is still fairly consistent.  Overall I think I like the simplicity of the current setup and building out the same API might be overengineering this.  

However, if we did want to change the API for consistency across the package I could imagine a .download() function saving a csv file of the table (which could be a simple pass-through to dataframe.to_csv()) & a .filter()  for somewhat who wanted to cut the table down based on say proper motion or magnitude if they realized that they grabbed too many sources in their initial query.

**Comment by christinahedges** [comment]
Based on comments at TSC3 it seems like people want to be able to put in a TIC/KIC/EPIC ID and get likely matches. We might do something like not giving people a cross-match exactly, but returning the likely hits from the catalog with a separation and relative flux so they could make their own cut.

We might need a facility to do this that can take in multiple TICIDs (see TessProposalTool...)

I think folks will want to get out extra keywords which we could add e.g. the stellar parameters.

**Comment by christinahedges** [comment]
I updated this PR to include stellar parameters as outputs and updated the tutorials. 

@tylerapritchard has pointed out that we could have an API like the one I described above e.g. 
```python
from lksearch.catalogs import TICSearch
c = SkyCoord.from_name('Kepler-10')
sr = TICSearch(c)
sr.table
sr.skycoords
```python

But I think this is overkill given what we want unless we add in some crossmatching functionality from the TESS proposal tool stuff. 

I think the piece missing to make this work is a more flexible interface like in `lksearch.MASTSearch`, e.g.

```python
query_TIC(f"{ra}, {dec}")
query_TIC((ra, dec))
```

and a way to resolve TIC, KIC, and EPIC names so we can do:

```python
query_TIC("TIC 3777809370")
query_gaia("TIC 3777809370")
query_EPIC("TIC 3777809370")
query_KIC("TIC 3777809370")
```

To do this we are going to need a way to query a given catalog by the ID and then return a SkyCoord object e.g.

```python
def TIC_to_SkyCoord(id:str) -> SkyCoord:
    ...
def KIC_to_SkyCoord(id:str) -> SkyCoord:
    ...
def EPIC_to_SkyCoord(id:str) -> SkyCoord:
    ...
def gaiadr3_to_SkyCoord(id:str) -> SkyCoord:
    ...
```

This should be a fast query that **isn't a radius query** to each catalog, because the unique ID name should give us a unique row without a radius query. We need to dig in to figure out how to do this in astroquery.

File: src/lksearch/CatalogSearch.py [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+    # Suppress warning caused by Astropy as noted in issue 111747 (https://github.com/astropy/astropy/issues/11747)
+    if epoch != equinox:
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", message="ERFA function")
+            warnings.filterwarnings("ignore", message="invalid value")
+            c = c.apply_space_motion(new_obstime=epoch)
+    return c
+
+
+class CatalogResult(Table):
```
Comment (christinahedges) [inline src/lksearch/CatalogSearch.py:614]: If we remove this, we can remove the functionality to turn a table into a sky coordinate, and then everything can be `pandas` dataframes. I think this is the best trade,  (we still have the functionality with return_skycoord=True) and we get to use pandas throughout the `lksearch` package outputs.

**Review by christinahedges — COMMENTED** [review]
@tylerapritchard This is a really awesome PR! There's lots of great functionality here. Other than the bug fixes we identified the main comments here are

1. You can remove the CatalogResult class and then have everything be in `pandas` dataframes. This will be important to keep the outputs consistent for users

2. The names need to follow Python style of having functions be lower case. I suggest we follow `verb_noun(noun)` style for function names and make the API something like this:

```python
from lksearch import MASTSearch, TESSSearch, KeplerSearch, K2Search
from lksearch.catalogs import query_region, query_ID, find_alternate_names, match_names_to_catalogs
query_region(region params)
query_id(ID)
find_alternate_names(name)
match_names_to_catalogs(name, match_strings=["tic", 'kic'])

```

**Pull request merged**